### PR TITLE
feature-flags: initial implementation

### PR DESCRIPTION
**Pull request by saghul**
See each commit message and RTFSC :-P

### CODE REVIEW

File: react/features/base/flags/actions.js [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+import { CLEAR_FLAGS, UPDATE_FLAGS } from './actionTypes';
+
+/**
+ * Clears all feature flags.
+ *
+ * @returns {{
+ *     type: CLEAR_FLAGS
+ * }}
+ */
+export function clearFlags() {
```
[the remark this answers is not in the document]
Comment (saghul) [reply react/features/base/flags/actions.js]: Nope.

[the remark this answers is not in the document]
Comment (paweldomas) [reply react/features/base/flags/actions.js]: then maybe it's not needed for now ?

**Comment by saghul** [comment]
@paweldomas Updated, PTAL!

File: ios/sdk/src/JitsiMeetConferenceOptions.m [hunk 2]
```diff
@@ -76,11 +89,14 @@ - (BOOL)videoMuted {
 }
 
 - (void)setWelcomePageEnabled:(BOOL)welcomePageEnabled {
-    _welcomePageEnabled = [NSNumber numberWithBool:welcomePageEnabled];
+    [self setFeatureFlag:WelcomePageEnabledFeastureFlag
```
Comment (paweldomas) [inline ios/sdk/src/JitsiMeetConferenceOptions.m]: a typo in `Feasture`

Comment (saghul) [reply ios/sdk/src/JitsiMeetConferenceOptions.m]: Stupid Xcode refactoring thing, didn't change all occurences! Fixed.

File: react/features/conference/components/native/Conference.js [hunk 3]
```diff
@@ -42,6 +42,13 @@ import type { AbstractProps } from '../AbstractConference';
  */
 type Props = AbstractProps & {
 
+    /**
+     * Wherther the calendar feasture is enabled or not.
```
Comment (paweldomas) [inline react/features/conference/components/native/Conference.js]: a typo in `feasture`
looks like your hand slips the same way whenever types this word :)

File: react/features/conference/components/native/Conference.js [hunk 4]
```diff
@@ -417,6 +424,14 @@ function _mapStateToProps(state) {
     return {
         ...abstractMapStateToProps(state),
 
+        /**
+         * Wherther the calendar feasture is enabled or not.
```
Comment (paweldomas) [inline react/features/conference/components/native/Conference.js]: a typo in `feasture`

File: react/features/conference/components/native/Conference.js (hunk 3 above)
Comment (saghul) [reply react/features/conference/components/native/Conference.js]: FML. I blame the new apple keyboard! :-P

File: ios/app/src/ViewController.m [hunk 5]
```diff
@@ -96,4 +96,10 @@ - (void)conferenceWillJoin:(NSDictionary *)data {
     [self _onJitsiMeetViewDelegateEvent:@"CONFERENCE_WILL_JOIN" withData:data];
 }
 
+#if 0
```
Comment (paweldomas) [inline ios/app/src/ViewController.m]: how does it work `#if 0` ?

Comment (saghul) [reply ios/app/src/ViewController.m]: it doesn't :-) it's there to make testing easier. just change it to 1 and one should see the change in behavior.

File: react/features/base/flags/reducer.js [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+ * @param {string} action.type - The type of the redux action to reduce.
+ * @returns {State} The next redux state that is the result of reducing the
+ * specified action.
+ */
+ReducerRegistry.register('features/base/flags', (state = DEFAULT_STATE, action) => {
+    switch (action.type) {
+    case UPDATE_FLAGS: {
+        const newState = _.merge({}, state, action.flags);
+
+        return _.isEqual(state, newState) ? state : newState;
```
Comment (paweldomas) [inline react/features/base/flags/reducer.js]: does it matter here to do the `isEqual` check here ? if it's equal then well... is still the same thing, no ?

Comment (saghul) [reply react/features/base/flags/reducer.js]: But the object would change, resulting in a state change? Maybe overly cautious, but hey.

File: react/features/toolbox/components/native/OverflowMenu.js [hunk 7]
```diff
@@ -29,6 +29,11 @@ type Props = {
      */
     _bottomSheetStyles: StyleType,
 
+    /**
+     * Whether the chat feature has been disabled. The meeting info button will be displayed in its place.
+     */
+    _chatDisabled: boolean,
```
Comment (paweldomas) [inline react/features/toolbox/components/native/OverflowMenu.js]: Can you please rename it to `_chatEnabled` ? It's a bit confusing going back and forth negating things. I mean here the `!` and `=== false` below on the props mapping.

File: react/features/toolbox/components/native/Toolbox.js [hunk 8]
```diff
@@ -202,7 +209,7 @@ class Toolbox extends Component<Props, State> {
      * @returns {React$Node}
      */
     _renderToolbar() {
-        const { _styles } = this.props;
+        const { _chatDisabled, _styles } = this.props;
```
Comment (paweldomas) [inline react/features/toolbox/components/native/Toolbox.js]: the same here, can we have it renamed to `_chatEnabled` ?

File: ios/app/src/ViewController.m (hunk 5 above)
Comment (paweldomas) [reply ios/app/src/ViewController.m]: then I don't get it, why is this code even here ?

File: react/features/base/flags/reducer.js (hunk 6 above)
Comment (paweldomas) [reply react/features/base/flags/reducer.js]: okay

File: ios/app/src/ViewController.m (hunk 5 above)
Comment (saghul) [reply ios/app/src/ViewController.m]: So we don't have to write it, eevery time it needs testing.

File: react/features/toolbox/components/native/OverflowMenu.js (hunk 7 above)
Comment (saghul) [reply react/features/toolbox/components/native/OverflowMenu.js]: Can do.

**Comment by saghul** [comment]
@paweldomas Updated again, PTAL!

**Comment by saghul** [comment]
jenkins please test this please

**Pull request merged**